sdk/go/pulumi-language-go/main.go

```diff
@@ -353,8 +353,14 @@ func execProgramCmd(cmd *exec.Cmd, env []string) error {
 		return errors.Wrapf(err, "program exited unexpectedly")
 	}
 
+	if status.ExitStatus() == 1 {
```

Comment:
Ugh, this does not feel right. Granted, up until recently Pulumi used `go run` therefore ExitStatus was always just either 1 or 2, with the real exit status printed to stderr, so existing programs don't rely on exit codes propagating out. Still. At least Pulumi used to always fail on non-0 exit codes from user code.

This can now swallow exit codes from unrelated conditions quite unexpectedly. 

Can we think of other options? Can we walk through what happens if we remove this and let the func return `return errors.Errorf("program exited with non-zero exit code: %d", status.ExitStatus())`?

If we cannot work this out cleanly, perhaps one crazy option is to overload the exit codes to use a <CUSTOM-INGEGER> reserved by Pulumi to indicate a special exit, that's unlikely to overlap with user code codes.

Reply:
I'm not sure I follow where RunResponse comes into play, but this is a blocker issue to me. Much more important not to swallow code 1 than to not print double error messages. What else can we do here?

Surfacing some discussion we had with @RobbieMcKinstry  also, what about having Run return exit code 0 if it reported the err to the engine? This seems to check out superficially, but you have more context on this:


```S1:

  - RunFunc returns non-nil err

    Context: func Run(body RunFunc, opts...RunOption)
   
    [[file:~/pulumi/sdk/go/pulumi/run.go::func Run(body RunFunc, opts ...RunOption) {]]

  - Run logs the error to the engine

  - Run finishes with exit code 0

  - pulumi-language-go finishes normally

  - but pulumi finishes with an error since error count != 0

  - only one error

S2:    

  - User code calls os.Exit(1) or panic() /// code=2 too, etc

  - Run does not even have a chance to run finalizers
    
  - pulumi-language-go finishes with exit code != 0

  - pulumi-language-go finishes with exit code != 0
      OR maybe it uses ctx.Log.Error to send it back to engine and finishes with exit code 0

  - only one error
```

Reply:
Upon discussion @dixler pointed out that Pulumi does not consider it an error when the engine receives several ERROR-level logs. So this approach does not work. @Frassle pointed out we might need protocol changes for the user program to indicate error exit. In the meanwhile we decided with @dixler to be pragmatic here and finish off this PR by using a dedicated non-1 pulumi-only exit code to indicate a failure that already has sent logs to the engine.